crates/ty_python_semantic/src/types.rs

```diff
@@ -9245,6 +9511,77 @@ impl<'db> BoundMethodType<'db> {
         )
     }
 
+    /// Infers this method scope's types and returns the inferred return type.
+    #[salsa::tracked(cycle_fn=return_type_cycle_recover, cycle_initial=return_type_cycle_initial, heap_size=get_size2::heap_size)]
```

Comment:
Does this need to be a separate Salsa query, or could we just infer the return type of `self.function(db)`?

We could add the union with `Unknown` here rather than inside `ScopeInference::infer_return_type`.